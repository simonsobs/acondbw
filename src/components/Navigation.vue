<template>
  <div class="navigation" style="position: relative;">
    <div v-if="state == State.LOADED">
      <v-list shaped>
        <v-list-item
          link
          router
          v-for="edge in edges"
          :key="edge.node.typeId"
          :to="'/' + edge.node.name"
        >
          <v-list-item-action>
            <v-icon v-text="edge.node.icon"></v-icon>
          </v-list-item-action>
          <v-list-item-content>
            <v-list-item-title v-text="edge.node.plural" class="capitalize"></v-list-item-title>
          </v-list-item-content>
        </v-list-item>
      </v-list>
    </div>
    <v-alert
      v-else-if="state == State.EMPTY"
      outlined
      dense
      type="info"
      class="ma-2"
    >No product types are defined.</v-alert>
    <div v-else-if="state == State.LOADING" class="mx-2 pt-5">
      <v-progress-circular indeterminate :size="26" color="grey"></v-progress-circular>
    </div>
    <v-alert v-else-if="state == State.ERROR" outlined dense type="error" class="ma-2">{{ error }}</v-alert>
    <dev-tool-loading-state-overriding-menu @state="devtoolState = $event"></dev-tool-loading-state-overriding-menu>
  </div>
</template>

<script>
import ALL_PRODUCTS_TYPES from "@/graphql/AllProductTypes.gql";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/DevToolLoadingStateOverridingMenu";

export default {
  name: "Navigation",
  components: {
    DevToolLoadingStateOverridingMenu
  },
  data: () => ({
    edges: null,
    error: null,
    devtoolState: null,
    State: State
  }),
  apollo: {
    edges: {
      query: ALL_PRODUCTS_TYPES,
      update: data =>
        data.allProductTypes ? data.allProductTypes.edges : null,
      result(result) {
        this.error = result.error ? result.error : null;
      }
    }
  },
  computed: {
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else if (this.edges) {
        if (this.edges.length) {
          return State.LOADED;
        } else {
          return State.EMPTY;
        }
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.$apollo.queries.edges.loading;
    }
  }
};
</script>
