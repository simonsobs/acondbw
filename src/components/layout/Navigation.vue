<template>
  <div class="navigation">
    <v-list v-if="loaded" shaped>
      <v-list-item
        link
        router
        v-for="edge in edges"
        :key="edge.node.typeId"
        :to="{
          name: 'ProductList',
          params: { productTypeName: edge.node.name },
        }"
      >
        <v-list-item-action class="mr-5">
          <v-icon v-text="edge.node.icon"></v-icon>
        </v-list-item-action>
        <v-list-item-content>
          <v-list-item-title
            v-text="edge.node.plural"
            class="capitalize condensed-font font-weight-medium"
          ></v-list-item-title>
        </v-list-item-content>
        <v-list-item-icon class="ml-2">
          <v-chip
            small
            v-if="!!edge.node.products.totalCount"
            v-text="edge.node.products.totalCount"
          ></v-chip>
        </v-list-item-icon>
      </v-list-item>
    </v-list>
    <v-alert v-else-if="empty" outlined dense type="info" class="ma-2">
      No product types are defined.
    </v-alert>
    <v-progress-circular
      v-else-if="loading"
      indeterminate
      :size="18"
      :width="3"
      color="secondary"
      class="mx-5 mt-5"
    ></v-progress-circular>
    <v-alert v-else-if="error" outlined dense type="error" class="ma-2">
      {{ error }}
    </v-alert>
    <v-bottom-navigation absolute class="px-3 justify-start align-center">
      <span class="grey--text text-caption">
        v{{ $store.getters.appVersion }}
      </span>
      <!-- <v-spacer></v-spacer>
      <v-icon>mdi-plus-thick</v-icon> -->
    </v-bottom-navigation>
    <dev-tool-loading-state-overriding-menu
      v-model="devtoolState"
    ></dev-tool-loading-state-overriding-menu>
  </div>
</template>

<script>
import ALL_PRODUCT_TYPES from "@/graphql/queries/AllProductTypes.gql";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/utils/DevToolLoadingStateOverridingMenu.vue";

export default {
  name: "Navigation",
  components: {
    DevToolLoadingStateOverridingMenu,
  },
  data: () => ({
    init: true,
    edges: null,
    error: null,
    devtoolState: null,
    State: State,
  }),
  computed: {
    state() {
      if (this.devtoolState) return this.devtoolState;
      if (this.$apollo.loading) return State.LOADING;
      if (this.error) return State.ERROR;
      if (this.edges) {
        if (this.edges.length) return State.LOADED;
        return State.EMPTY;
      }
      if (this.init) return State.INIT;
      return State.NONE;
    },
    loading() {
      return this.state == State.LOADING;
    },
    loaded() {
      return this.state == State.LOADED;
    },
    empty() {
      return this.state == State.EMPTY;
    },
    notFound() {
      return this.state == State.NONE;
    },
  },
  watch: {
    devtoolState: function () {
      if (this.devtoolState) {
        this.init = this.devtoolState == State.INIT;
      }
      this.error =
        this.devtoolState == State.ERROR ? "Error from Dev Tools" : null;
    },
    "$store.state.nApolloMutations": function () {
      this.$apollo.queries.edges.refetch();
    },
  },
  apollo: {
    edges: {
      query: ALL_PRODUCT_TYPES,
      update: (data) =>
        data.allProductTypes ? data.allProductTypes.edges : null,
      result(result) {
        this.init = false;
        this.error = result.error || null;
      },
    },
  },
};
</script>
