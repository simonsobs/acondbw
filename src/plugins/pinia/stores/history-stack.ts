import { computed } from "vue";
import { defineStore } from "pinia";
import {
  useRouter,
  RouteLocationRaw,
  RouteLocationNormalized,
} from "vue-router";
import { useSessionStorage } from "@vueuse/core";

export const useHistoryStack = defineStore("historyStack", () => {
  const historyStack = useSessionStorage<RouteLocationRaw[]>(
    "historyStack",
    []
  );
  const router = useRouter();

  const afterEach = (
    to: RouteLocationNormalized,
    from: RouteLocationNormalized
  ) => {
    if (to.name !== "ProductItem") {
      historyStack.value.splice(0); // clear
      return;
    }

    if (!historyStack.value.length && from.name === "ProductList") {
      // push "from" so that we can move back to "list"
      const pushArg = toPushArg(from);
      historyStack.value.push(pushArg);
    }

    const pushArg = toPushArg(to);
    if (isSamePathAsLast(pushArg)) return; // Always true while moving back.

    historyStack.value.push(pushArg);
  };

  /**
   * Return true if the same as the last item in the history stack
   */
  function isSamePathAsLast(arg: RouteLocationRaw) {
    if (!historyStack.value.length) return false;
    return isSamePath(historyStack.value[historyStack.value.length - 1], arg);
  }

  /**
   * Return true if two routes resolved to the same path
   *
   * @see https://router.vuejs.org/guide/essentials/route-matching-syntax.html
   */
  function isSamePath(r1: RouteLocationRaw, r2: RouteLocationRaw) {
    return router.resolve(r1).href === router.resolve(r2).href;
  }

  /**
   * Convert an argument of afterEach() to an argument of router.push()
   *
   * @see https://router.vuejs.org/guide/essentials/navigation.html
   */
  function toPushArg(arg: RouteLocationNormalized): RouteLocationRaw {
    return { path: arg.path };
  }

  const hasBack = computed(() => historyStack.value.length >= 2);

  function back() {
    if (!hasBack.value) return;
    historyStack.value.pop();
    router.push(historyStack.value[historyStack.value.length - 1]);
  }

  return {
    historyStack,
    afterEach,
    back,
    hasBack,
  };
});
