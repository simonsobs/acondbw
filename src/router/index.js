import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
