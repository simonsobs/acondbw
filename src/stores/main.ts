import { ref, computed } from "vue";
import { defineStore } from "pinia";

export const useStore = defineStore("main", () => {
  const example = ref("abc");
  const snackbar = ref(false);
  const snackbarMessage = ref<string | null>(null);
  const nApolloMutations = ref(0);
  const packageVersion = ref(import.meta.env.PACKAGE_VERSION || "vx.x.x");
  const appVersion = computed(() => packageVersion.value);
  function loadExample() {
    example.value = "123";
  }
  function setSnackbarMessage(message: string) {
    snackbarMessage.value = message;
    snackbar.value = true;
  }
  function closeSnackbar() {
    snackbar.value = false;
  }
  function apolloMutationCalled() {
    nApolloMutations.value++;
  }
  return {
    example,
    snackbar,
    snackbarMessage,
    nApolloMutations,
    packageVersion,
    appVersion,
    loadExample,
    setSnackbarMessage,
    closeSnackbar,
    apolloMutationCalled,
  };
});
