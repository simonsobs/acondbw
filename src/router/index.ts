import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";
import VueMeta from "vue-meta";

import { default as defaultPinia } from "@/stores";
import { useAuthStore } from "@/stores/auth";

import Frame from "@/components/layout/Frame.vue";
import NullFrame from "@/components/layout/NullFrame.vue";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";
import SearchTest from "@/views/framework/SearchTest.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue";

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import ProductType from "@/views/admin/ProductType.vue";

const About = () => import("@/views/framework/About.vue");
const NotFound = () => import("@/views/framework/NotFound.vue");
const AccessDenied = () => import("@/views/framework/AccessDenied.vue");

const FrameAdmin = () =>
  import(/* webpackChunkName: "admin" */ "@/components/admin/FrameAdmin.vue");
const Version = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Version.vue");
const Log = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Log.vue");
const Config = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Config.vue");
const Theme = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Theme.vue");
const User = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/User.vue");
const AdminAppToken = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken.vue"
  );
const AdminAppAuth = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth.vue"
  );
const AdminAppTokenError = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError.vue"
  );

Vue.use(VueRouter);
Vue.use(VueMeta);

let pinia = defaultPinia;

function setPinia(val) {
  // because not sure how to monkey patch pinia for tests
  pinia = val;
}

function setDefaultPinia() {
  setPinia(defaultPinia);
}

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    // meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/search-test",
    name: "SearchTest",
    components: { default: SearchTest, frame: Frame },
    meta: { requiresAuth: true },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: { default: OAuthRedirect, frame: NullFrame },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: { default: SignIn, frame: NullFrame },
  },
  {
    path: "/auth",
    name: "Auth",
    components: { default: Auth, frame: NullFrame },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: { default: SignInError, frame: NullFrame },
  },
  {
    path: "/admin",
    name: "Admin",
    redirect: { name: "AdminVersion" },
  },
  {
    path: "/admin/product-type",
    name: "ProductType",
    components: { default: ProductType, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/versions",
    name: "AdminVersion",
    components: { default: Version, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/log",
    name: "AdminLog",
    components: { default: Log, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/config",
    name: "AdminConfig",
    components: { default: Config, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/theme",
    name: "AdminTheme",
    components: { default: Theme, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/users",
    name: "AdminUser",
    components: { default: User, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: { default: AdminAppToken, frame: FrameAdmin },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: { default: AdminAppAuth, frame: FrameAdmin },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: { default: AdminAppTokenError, frame: FrameAdmin },
  },
  {
    path: "/about",
    name: "About",
    component: About,
  },
  {
    path: "/add-product/:productTypeName",
    name: "ProductAdd",
    components: { default: ProductAdd, frame: NullFrame },
    meta: { requiresAuth: true },
  },
  {
    path: "/product/:productTypeName",
    components: { default: ProductTop, frame: Frame },
    meta: { requiresAuth: true },
    children: [
      { path: "", name: "ProductList", component: ProductList },
      { path: ":name", name: "ProductItem", component: ProductItem },
    ],
  },
  {
    path: "/access-denied",
    name: "AccessDenied",
    components: { default: AccessDenied, frame: NullFrame },
  },
  {
    path: "*",
    name: "NotFound",
    components: { default: NotFound, frame: NullFrame },
  },
];

function createRouter() {
  const router = new VueRouter({
    mode: "history",
    base: process.env.BASE_URL,
    routes,
  });

  router.beforeEach((to, from, next) => {
    const auth = useAuthStore(pinia);

    const adminRequired = to.matched.some(
      (record) => record.meta.requiresAdmin
    );
    const isAdmin = auth.isAdmin;
    if (adminRequired && !isAdmin) {
      next({ name: "AccessDenied" });
      return;
    }

    const authRequired = to.matched.some((record) => record.meta.requiresAuth);
    const isSignedIn = auth.isSignedIn as boolean;
    if (authRequired && !isSignedIn) {
      next({ name: "AccessDenied" });
      return;
    }
    next();
  });

  return router;
}

const router = createRouter();

function checkAuthForCurrentRoute() {
  const authRequired = router.currentRoute.matched.some(
    (record) => record.meta.requiresAuth
  );
  const auth = useAuthStore(pinia);
  const signedIn = auth.isSignedIn as boolean;
  if (authRequired && !signedIn) {
    router.push("/");
  }
}

export { createRouter, checkAuthForCurrentRoute, setPinia, setDefaultPinia };
