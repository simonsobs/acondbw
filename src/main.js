import Vue from "vue";
import App from "./App.vue";
import router from "./router";
import store from "./store";
import vuetify from "./plugins/vuetify";
import { createProvider, AUTH_TOKEN } from "./vue-apollo";
import { setContext } from "apollo-link-context";

Vue.config.productionTip = false;

// copied from https://blog.logrocket.com/handling-authentication-in-your-graphql-powered-vue-app/
const authLink = setContext(async (_, { headers }) => {
  // get the authentication token from local storage if it exists
  const token = JSON.parse(localStorage.getItem(AUTH_TOKEN));
  // const token = localStorage.getItem(AUTH_TOKEN);
  // Return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      authorization: token || "",
    },
  };
});

console.log(authLink)
const options = { link: authLink };

new Vue({
  router,
  vuetify,
  store,
  apolloProvider: createProvider(options),
  render: (h) => h(App),
  created: function() {
    this.$store.dispatch("setTokenFromLocalStorage");
  }
}).$mount("#app");
