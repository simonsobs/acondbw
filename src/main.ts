import { createApp, h } from "vue";

import pinia from "@/stores";
import { createRouter_ } from "@/router";
import vuetify from "@/plugins/vuetify";
import App from "@/app/App.vue";
import DevToolLoadingStateMenu from "@/components/utils/DevToolLoadingStateMenu.vue";

const router = createRouter_();

const app = createApp({
  render: () => h(App),
});
app.component("DevToolLoadingStateMenu", DevToolLoadingStateMenu);
app.use(pinia);
app.use(router);
app.use(vuetify);
app.mount("#app");
