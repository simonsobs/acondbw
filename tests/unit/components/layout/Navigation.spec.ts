import Vue from "vue";
import VueRouter from "vue-router";
import { PiniaVuePlugin } from "pinia";
import Vuetify from "vuetify";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";

import Navigation from "@/components/layout/Navigation.vue";
import { createRouter } from "@/router";
import { useStore } from "@/stores/main";

jest.mock("vue-apollo");
// To prevent the error: "[vue-test-utils]: could not overwrite
// property $apollo, this is usually caused by a plugin that has added
// the property as a read-only value"
// https://github.com/vuejs/vue-apollo/issues/798

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("App.vue", () => {
  let localVue: ReturnType<typeof createLocalVue>;
  let vuetify: Vuetify;
  let router: ReturnType<typeof createRouter>;
  let wrapper: ReturnType<typeof shallowMount>;
  let store: ReturnType<typeof useStore>;

  const edges = [
    {
      node: {
        id: "UHJvZHVjdFR5cGU6Mw==",
        typeId: "3",
        name: "simulation",
        order: 1,
        indefArticle: "a",
        singular: "simulation",
        plural: "simulations",
        icon: "mdi-creation",
        products: {
          totalCount: 0,
        },
      },
    },
    {
      node: {
        id: "UHJvZHVjdFR5cGU6MQ==",
        typeId: "1",
        name: "map",
        order: 2,
        indefArticle: "a",
        singular: "map",
        plural: "maps",
        icon: "mdi-map",
        products: {
          totalCount: 58,
        },
      },
    },
    {
      node: {
        id: "UHJvZHVjdFR5cGU6Mg==",
        typeId: "2",
        name: "beam",
        order: 3,
        indefArticle: "a",
        singular: "beam",
        plural: "beams",
        icon: "mdi-spotlight-beam",
        products: {
          totalCount: 5,
        },
      },
    },
  ];

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(PiniaVuePlugin);
    vuetify = new Vuetify();
    router = createRouter();
    const pinia = createTestingPinia();
    wrapper = shallowMount(Navigation, {
      localVue,
      vuetify,
      pinia,
      router,
      mocks: {
        $apollo: {
          loading: false,
          queries: {
            edges: {
              loading: false,
            },
          },
        },
      },
    });
    wrapper.setData({
      edges: edges,
    });
    store = useStore(pinia);
    store.packageVersion = "0.1.1";
  });

  it("match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
