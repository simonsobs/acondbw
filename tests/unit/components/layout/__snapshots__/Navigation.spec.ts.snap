// Vitest Snapshot v1

exports[`App.vue > match snapshot 1`] = `
"<div class=\\"navigation\\">
  <vuecomponent-stub loaderheight=\\"4\\" tag=\\"div\\" flat=\\"true\\">
    <vuecomponent-stub shaped=\\"true\\" tag=\\"div\\">
      <vuecomponent-stub activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
        <vuecomponent-stub class=\\"mr-5\\">
          <vuecomponent-stub>mdi-creation</vuecomponent-stub>
        </vuecomponent-stub>
        <vuecomponent-stub tag=\\"div\\">
          <vuecomponent-stub tag=\\"div\\" class=\\"capitalize condensed-font font-weight-medium\\">simulations</vuecomponent-stub>
        </vuecomponent-stub>
        <vuecomponent-stub class=\\"ml-2\\">
          <!---->
        </vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
        <vuecomponent-stub class=\\"mr-5\\">
          <vuecomponent-stub>mdi-map</vuecomponent-stub>
        </vuecomponent-stub>
        <vuecomponent-stub tag=\\"div\\">
          <vuecomponent-stub tag=\\"div\\" class=\\"capitalize condensed-font font-weight-medium\\">maps</vuecomponent-stub>
        </vuecomponent-stub>
        <vuecomponent-stub class=\\"ml-2\\">
          <vuecomponent-stub small=\\"true\\" activeclass=\\"\\" tag=\\"span\\" active=\\"true\\" closeicon=\\"$delete\\" closelabel=\\"$vuetify.close\\" filtericon=\\"$complete\\">58</vuecomponent-stub>
        </vuecomponent-stub>
      </vuecomponent-stub>
      <vuecomponent-stub activeclass=\\"\\" link=\\"true\\" to=\\"[object Object]\\" tag=\\"div\\" router=\\"\\">
        <vuecomponent-stub class=\\"mr-5\\">
          <vuecomponent-stub>mdi-spotlight-beam</vuecomponent-stub>
        </vuecomponent-stub>
        <vuecomponent-stub tag=\\"div\\">
          <vuecomponent-stub tag=\\"div\\" class=\\"capitalize condensed-font font-weight-medium\\">beams</vuecomponent-stub>
        </vuecomponent-stub>
        <vuecomponent-stub class=\\"ml-2\\">
          <vuecomponent-stub small=\\"true\\" activeclass=\\"\\" tag=\\"span\\" active=\\"true\\" closeicon=\\"$delete\\" closelabel=\\"$vuetify.close\\" filtericon=\\"$complete\\">5</vuecomponent-stub>
        </vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
    <!---->
    <vuecomponent-stub tag=\\"div\\" class=\\"px-5 py-0\\">
      <vuecomponent-stub tag=\\"div\\"></vuecomponent-stub>
      <vuecomponent-stub opendelay=\\"800\\" closedelay=\\"0\\" left=\\"true\\" openonclick=\\"true\\" openonhover=\\"true\\" openonfocus=\\"true\\" contentclass=\\"\\" maxwidth=\\"auto\\" nudgebottom=\\"0\\" nudgeleft=\\"0\\" nudgeright=\\"0\\" nudgetop=\\"0\\" nudgewidth=\\"0\\" tag=\\"span\\"><span> Add a product type </span></vuecomponent-stub>
    </vuecomponent-stub>
    <!---->
    <!---->
  </vuecomponent-stub>
  <vuecomponent-stub absolute=\\"true\\" height=\\"56\\" inputvalue=\\"true\\" activeclass=\\"v-btn--active\\" tag=\\"div\\" class=\\"px-3 justify-start align-center\\"><span class=\\"grey--text text-caption\\"> v0.1.1 </span></vuecomponent-stub>
  <devtoolloadingstatemenu-stub top=\\"-10px\\" right=\\"-10px\\"></devtoolloadingstatemenu-stub>
</div>"
`;
